### DISCUSSION TITLE
Can you share the decoding and sharing of the original data stream of scrcpy?

### DISCUSSION THREAD

**Discussion by sbfkcel** [yume-chan/ya-webadb#381] [General]
I try to parse scrcpy raw data stream in web.

But it's always not perfect and can only parse a few frames. Can you share your thoughts?

**Comment by yume-chan** [comment]
Because you can at least decoding several frames, I don't think there is any major issue with your stream parsing code.

If you are using ADB to communicate with scrcpy, then maybe you accidentally blocked the stream: ADB is a very delicate multiplex protocol, depending on the exact method you use (and their internal buffers), not reading all outputs from all sockets, or sometime even reading in wrong order, will completely hang all the streams. For example I have:

https://github.com/yume-chan/ya-webadb/blob/4d0f1a11cb364414fab75d05ba52f63a1dc790fc/libraries/scrcpy/src/connection.ts#L115-L118

You can verify this by logging incoming data, and checking whether it stops.

**Comment by yume-chan** [comment]
My scrcpy client implementation is in [libraries/scrcpy/src](https://github.com/yume-chan/ya-webadb/tree/eded5a639ea25c62fb14a6e449d1ff8c894b3f67/libraries/scrcpy/src) folder. 

`connection.ts` has connection logic, `client.ts` and `options/1_16/index.ts` are video stream parsing related.

Scrcpy 1.22 introduced 2 new hidden options affecting stream format (#374). They are not available unless you directly start scrcpy server (not using scrcpy client).

If you don't care about them, you can go back to 42c7da9350ed41218b31d52275f84fde54ff3fac, the code is much simpler without branches for those options.

**Comment by sbfkcel** [comment]
good. Thank you so much.

I think it will take me some time to understand...

According to the official documentation of scrcpy, when the rotation direction of the device changes, a new data stream will be generated. What is the data differentiation logic here (that is, how can the rotation direction be changed through the buffer data)?

**Reply by yume-chan** [reply]
As I mentioned in #374 

> However, the video socket is not always **one** H.264 stream. When device screen rotates, video size will change, and a new H.264 stream with new size replaces the old one.
>
> Checking `pts == 0` can detect such stream restarts, parse video encoding/size, and re-initialize decoder. In theory, I can also parse the whole stream to find restarts, but that will be super complicate (I need to implement all H.264 protocol) and inefficient (parse whole stream twice).

**Reply by sbfkcel** [reply]
The width and height can be obtained by decoding the video size, and only vertical or horizontal can be distinguished. Is there a good way to determine whether to rotate to the left or to the right?

**Reply by yume-chan** [reply]
Scrcpy doesn't tell you that.

You can use other methods, for example https://stackoverflow.com/a/26955141/1179747

**Reply by sbfkcel** [reply]
cool

**Comment by sbfkcel** [comment]
thanks.

Another question, sending the get clipboard content command. Is the data here returning the clipboard content from the control connection? Still mixed with the video data stream. Please try to verify, the implementation section was not found in the code.

**Reply by yume-chan** [reply]
get clipboard request is new in 1.22, I haven't added support for that.

**Comment by sbfkcel** [comment]
Can scrcpy dynamically set instructions for stream-related parameters (such as modifying bit rate, video size) after the stream has been sent?

**Reply by yume-chan** [reply]
No.
